py: Add option to cache LOAD_NAME/LOAD_GLOBAL result in bytecode.

dpgeorge opened the pull request:
This is a simple optimisation inspired by JITing technology: we cache in the bytecode (using 1 byte) the offset of the last successful lookup in dict_locals or dict_globals.  This allows us to next time round check in that location in the hash table (mp_map_t) for the desired entry, and if it's there, use that entry straight away.  Otherwise fallback to a normal map lookup.

On a few tests it gives >90% cache hit and greatly improves speed of code relying on globals.

Pystone test results run 6 times with this new optimisation:

```
Pystone(1.1) time for 200000 passes = 2.559 This machine benchmarks at 78155.5 pystones/second
Pystone(1.1) time for 200000 passes = 2.711 This machine benchmarks at 73773.5 pystones/second
Pystone(1.1) time for 200000 passes = 2.644 This machine benchmarks at 75643 pystones/second
Pystone(1.1) time for 200000 passes = 2.596 This machine benchmarks at 77041.6 pystones/second
Pystone(1.1) time for 200000 passes = 2.497 This machine benchmarks at 80096.1 pystones/second
Pystone(1.1) time for 200000 passes = 2.558 This machine benchmarks at 78186.1 pystones/second
```

Results without optimisation:

```
Pystone(1.1) time for 200000 passes = 3.061 This machine benchmarks at 65338.1 pystones/second
Pystone(1.1) time for 200000 passes = 3.117 This machine benchmarks at 64164.3 pystones/second
Pystone(1.1) time for 200000 passes = 3.034 This machine benchmarks at 65919.6 pystones/second
Pystone(1.1) time for 200000 passes = 3.026 This machine benchmarks at 66093.9 pystones/second
Pystone(1.1) time for 200000 passes = 3.428 This machine benchmarks at 58343.1 pystones/second
Pystone(1.1) time for 200000 passes = 3.237 This machine benchmarks at 61785.6 pystones/second
```

Pystone varies a lot, but we see here about 0.5s reduction in running time (3.1 down to 2.6s) which is a 16% reduction in running time, and about 16% increase in pystones.

dpgeorge committed:
py: Add option to cache LOAD_NAME/LOAD_GLOBAL result in bytecode.

dpgeorge commented:
Note that above tests were run with standard `micropython` (ie not `micropython_fast`) on 32-bit x86 machine.

dpgeorge commented:
With this optimisation enabled, code size increases by 276 bytes on x86 unix and 196 bytes on stmhal.  RAM usage will increase by 1 byte per LOAD_NAME/LOAD_GLOBAL opcode.  I think these are fair increases given the speed improvement.

dpgeorge commented:
Following this would be trying to do the same thing for LOAD_ATTR and LOAD_METHOD, which are a bit trickier.

dpgeorge committed:
py: Add option to cache LOAD_ATTR result in bytecode.

dpgeorge commented:
Added further option to cache LOAD_ATTR results.  This improves lookup of attributes in user-defined instances.

pfalcon commented:
I read about such optimizations - they say Smalltalk largely relied on it, to pay off "everything is object and message send" approach. When I tried to map them to to uPy, I thought "wait, but uPy''s requirement is to have very compact bytecode, that's why we don't use word-sized bytecode, etc. Then, we don't want to 4 bytes for inline cache here and there - instead, all optimizations would rather go to native codegeneration." But spending 1 byte on that is surely a nice trick ;-).

But 15% speedup on pystone surely looks nice, but non-conclusive. Did you try to create micro-benchmark to show some 100% speedup?

Patch looks good (my guess is that starting with a single config define would be good enough, unless you have special usecase).

dpgeorge committed:
py: Add option to cache STORE_ATTR result in bytecode.

dpgeorge commented:
Also implemented caching trick for STORE_ATTR, which can make a big difference as well.

dpgeorge commented:
Some results for Pystone(1.2) with 200000 iterations on my x64 machine (number given is approx number of pystones averaged by eye over a few runs):
- CPython3: 70k
- uPy std, no caching: 60k
- uPy std, all caching options enabled: 90k
- uPy fast, no caching: 80k
- uPy fast, all caching options enabled: 115k !!

That's pretty nice.

dpgeorge commented:
> But spending 1 byte on that is surely a nice trick ;-).

Yes, it works out very neatly: 1 byte is plenty for most dicts (especially instance objects, and probably global dict), and if the map is more than 255 entries long, the cache will still work for those entries in positions <= 255, and slow path for others.  And there's also no need to have a dynamic code counter to work out hotspots or anything, you just optimise all loads/stores because it's cheap to do.  With cache miss it's also not that bad because the check is so cheap.

dpgeorge commented:
And you don't need to keep a hash of the type/dict around or anything to check if the map changed.  If you find the entry at the guessed byte-offset you are good to go.

dpgeorge commented:
> But 15% speedup on pystone surely looks nice, but non-conclusive.

See new results: for uPy fast (make -B fast) it's close to a 50% increase in pystones.

> Did you try to create micro-benchmark to show some 100% speedup?

Best I did was run my standard triple-loop test which uses globals for all variables, and it goes from 15s to 5s running time.

pfalcon commented:
Sounds great!

dpgeorge committed:
py: Make MICROPY_BYTECODE_CACHE_STORE_ATTR work with del and property.

dpgeorge commented:
Cleaned up, rebased and merged in 7ee91cf86136f4774a9143a3f745d37f314662da.

With this optimisation enabled, code increases by:
- unix x64 up by 528 bytes
- stmhal up by 460 bytes
